File: prextras-agent/src/main/java/com/nicologies/prextras/BuildService.java
package com.nicologies.prextras;

import com.nicologies.prextras.common.PrExtrasConstants;
import com.nicologies.prextras.common.SettingsKeys;
import jetbrains.buildServer.RunBuildException;
import jetbrains.buildServer.agent.AgentRunningBuild;
import jetbrains.buildServer.agent.runner.BuildServiceAdapter;
import jetbrains.buildServer.agent.runner.ProgramCommandLine;
import jetbrains.buildServer.util.StringUtil;
import org.eclipse.egit.github.core.*;
import org.eclipse.egit.github.core.service.GitHubService;
import org.eclipse.egit.github.core.service.IssueService;
import org.eclipse.egit.github.core.service.PullRequestService;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.security.InvalidParameterException;
import java.util.*;

public class BuildService extends BuildServiceAdapter {
    private boolean _isOnWindows = true;
    private static final String BranchNameParamName = "teamcity.build.pull_req.branch_name";
    public BuildService() {
    }

    @Override
    public void beforeProcessStarted() throws RunBuildException {
        getLogger().progressMessage("Running PrExtras");
        Map<String,String> runnerParams = getRunnerParameters();
        Map<String,String> configParams = getConfigParameters();

        _isOnWindows = configParams.get("teamcity.agent.jvm.os.name").toLowerCase().startsWith("win");

        try {
            AgentRunningBuild build = getBuild();

            PullRequestService service = new PullRequestService();
            initService(service);
            IssueService issueService = new IssueService(service.getClient());

            String prNum = configParams.get("teamcity.build.branch");
            PullRequest pullRequest = null;
            boolean isPullRequestBuild = configParams.containsValue("refs/pull/" + prNum + "/merge")
                    || configParams.containsValue("refs/pull/" + prNum + "/head");
            isPullRequestBuild &= StringUtil.isNumber(prNum);
            RepositoryId repo = getRepository(configParams);
            if(isPullRequestBuild){
                pullRequest = getPullRequest(service, repo, prNum);
            }
            build.addSharedConfigParameter("teamcity.build.pull_req.is_pull_req",
                    isPullRequestBuild? "true":"false");

            String branchName = getBranchName(pullRequest, prNum);

            build.addSharedConfigParameter(BranchNameParamName, branchName);

            ExportPullRequestExtraInfo(pullRequest, build, configParams, issueService);

            if(pullRequest != null && pullRequest.getComments() > 0){
                ExportPrParticipants(service, repo, pullRequest.getNumber());
            }

            String appendToBuildNum = runnerParams.get(SettingsKeys.AppendToBuildNum);
            if(!StringUtil.isEmptyOrSpaces(appendToBuildNum)
                    && appendToBuildNum.toLowerCase().equals("true")) {
                String buildNum = getConfigParameters().get("build.number");
                boolean isGitVersionBuildNum = buildNum.matches("(?i).*PullRequest\\.\\d*\\+.*");
                if(isGitVersionBuildNum){
                    buildNum = buildNum.replaceAll("(?i)PullRequest\\.\\d*", branchName);
                }else if(!buildNum.contains(branchName)){
                    buildNum = buildNum + "-" + branchName;
                }
                buildNum = buildNum.replaceAll("(?i)PullRequest", "PR");
                String param = "##teamcity[buildNumber " + "'" + buildNum + "']";
                getLogger().message(param);
            }

            FailBuildIfConflict(runnerParams, pullRequest);
        } catch(RunBuildException e){
            throw e;
        }
        catch (Exception e) {
            throw new RunBuildException(e.getMessage(), e);
        }
    }

    private void ExportPullRequestExtraInfo(PullRequest pullRequest,
                                            AgentRunningBuild build,
                                            Map<String, String> configParams,
                                            IssueService issueService) {
        if(pullRequest == null){
            build.addSharedConfigParameter("teamcity.build.pull_req.is_merged", "false");
            build.addSharedConfigParameter("teamcity.build.pull_req.is_closed", "false");
            return;
        }
        build.addSharedConfigParameter("teamcity.build.pull_req.is_merged", pullRequest.isMerged()? "true" : "false");
        build.addSharedConfigParameter("teamcity.build.pull_req.is_closed",
                pullRequest.getClosedAt() != null? "true" : "false");

        User user = pullRequest.getUser();
        String email = user.getEmail();
        if(!StringUtil.isEmptyOrSpaces(email)) {
            build.addSharedConfigParameter("teamcity.build.pull_req.author_email", email);
        }

        String authorName = getGitHubUser(user);
        if(!StringUtil.isEmptyOrSpaces(authorName)){
            build.addSharedConfigParameter("teamcity.build.pull_req.author", authorName);
        }

        String htmlUrl = pullRequest.getHtmlUrl();
        if(!StringUtil.isEmptyOrSpaces(htmlUrl)) {
            build.addSharedConfigParameter("teamcity.build.pull_req.url", htmlUrl);
        }

        User assignee = pullRequest.getAssignee();
        if(assignee != null){
            String assigneeName = getGitHubUser(assignee);
            if(!StringUtil.isEmptyOrSpaces(assigneeName)) {
                build.addSharedConfigParameter("teamcity.build.pull_req.assignee", assigneeName);
            }
            email = assignee.getEmail();
            if(!StringUtil.isEmptyOrSpaces(email)) {
                build.addSharedConfigParameter("teamcity.build.pull_req.assignee_email", email);
            }
        }
        RepositoryId repo = getRepository(configParams);
        try {
            Issue issue = issueService.getIssue(repo, pullRequest.getNumber());
            List<Label> labels = issue.getLabels();
            if(!labels.isEmpty()) {
                String labelsString = StringUtil.join(";", labels.toArray());
                build.addSharedConfigParameter("teamcity.build.pull_req.labels", labelsString);
            }
        } catch (IOException e) {
            getLogger().message(e.getMessage() + e.getStackTrace());
            throw new RuntimeException(e.getMessage(), e);
        }
    }

    private void ExportPrParticipants(PullRequestService service, RepositoryId repo, int prNum) {
        try {
            HashSet<String> participants = new HashSet<String>();
            getParticipantsFromComments(participants, service.getComments(repo, prNum));
            IssueService issueService = new IssueService();
            initService(issueService);
            getParticipantsFromComments(participants, issueService.getComments(repo, prNum));
            if(participants.isEmpty()){
                return;
            }

            getBuild().addSharedConfigParameter("teamcity.build.pull_req.participants",
                    StringUtil.join(";", participants));
        } catch (IOException e) {
            e.printStackTrace();
            getLogger().error(e.getMessage());
        }
    }

    private void getParticipantsFromComments(HashSet<String> participants, List<? extends Comment> comments) {
        for(Comment comment : comments){
            String user = getGitHubUser(comment.getUser());
            if(StringUtil.isEmptyOrSpaces(user)){
                continue;
            }
            participants.add(user);
        }
    }

    private static String getGitHubUser(User user){
        if(user == null){
            return null;
        }
        String login = user.getLogin();
        if(StringUtil.isEmptyOrSpaces(login)){
            return null;
        }
        return login;
    }

    private void FailBuildIfConflict(Map<String, String> runnerParams,
                                     PullRequest pr) throws RunBuildException {
        if(pr == null){
            return;
        }
        String failBuildIfConflict = runnerParams.get(SettingsKeys.FailBuildIfConflict);

        if(!StringUtil.isEmptyOrSpaces(failBuildIfConflict) && failBuildIfConflict.equalsIgnoreCase("true")){
            if(!pr.isMerged() && !pr.isMergeable()){
                String msg = "The pull request has conflicts " + pr.getHtmlUrl();
                throw new RunBuildException(msg);
            }
        }
    }

    private String getBranchName(PullRequest pr, String prNum) throws RunBuildException{
        if(!StringUtil.isNumber(prNum)) {
            return prNum;
        }
        try {
            return pr != null? pr.getHead().getRef(): prNum;
        }catch (Exception ex){
            throw new RunBuildException("Unable to get branch name for pull request " + ex.getMessage(), ex);
        }
    }

    private PullRequest getPullRequest(PullRequestService service, RepositoryId repo, String prNum)
            throws RunBuildException {
        PullRequest pr;
        int prNumInteger;
        try {
            prNumInteger = Integer.parseInt(prNum);
        } catch (Exception e) {
            throw new RunBuildException(prNum + " is not pull request number: " + e.getMessage() , e);
        }

        getLogger().message("Trying to get extra information of pull request " + prNum + " from "
                + repo.getOwner() + "'s " + repo.getName());
        try {
            pr = service.getPullRequest(repo, prNumInteger);
        } catch (IOException e) {
            throw new RunBuildException("unable to get pull request info: " + e.getMessage(), e);
        }
        return pr;
    }

    private RepositoryId getRepository(Map<String, String> configParams) {
        String repoUrl = configParams.get("vcsroot.url");
        return RepoInfoParser.Parse(repoUrl);
    }

    @NotNull
    private void initService(GitHubService service) {
        Map<String, String> runnerParams = getRunnerParameters();
        String authType = runnerParams.get(SettingsKeys.AuthType);
        if(authType.equals(PrExtrasConstants.SystemWideTokenAuthType)){
            String token = getSystemProperties().get(SettingsKeys.GithubToken);
            if(StringUtil.isEmptyOrSpaces(token)){
                throw new InvalidParameterException("Please set the system wide github token");
            }
            service.getClient().setOAuth2Token(token);
        }
        else if(authType.equals(PrExtrasConstants.TokenAuthType)){
            String token = runnerParams.get(SettingsKeys.GithubToken);
            if(StringUtil.isEmptyOrSpaces(token)){
                throw new InvalidParameterException("Please set the github token in the build step's settings page");
            }
            service.getClient().setOAuth2Token(token);
        }
        else{
            String password = runnerParams.get(SettingsKeys.GithubPassword);
            if(StringUtil.isEmptyOrSpaces(password)){
                password = null;
            }
            String user = runnerParams.get(SettingsKeys.GithubUserName);
            if(StringUtil.isNotEmpty(user)) {
                service.getClient().setCredentials(user, password);
            }
        }
    }

    @NotNull
    public ProgramCommandLine makeProgramCommandLine() throws RunBuildException {
        return new ProgramCommandLine() {
            @NotNull
            public String getExecutablePath() throws RunBuildException {
                return _isOnWindows? "cmd" : "sh";
            }

            @NotNull
            public String getWorkingDirectory() throws RunBuildException {
                return getCheckoutDirectory().getPath();
            }

            @NotNull
            public List<String> getArguments() throws RunBuildException {
                List<String > ret = new Vector<String>();
                if(_isOnWindows)
                {
                    ret.add("/c");
                }
                ret.add("echo");
                ret.add("Branch name is " + getConfigParameters().get(BranchNameParamName));
                return ret;
            }

            @NotNull
            public Map<String, String> getEnvironment() throws RunBuildException {
                return getBuildParameters().getEnvironmentVariables();
            }
        };
    }
}

File: prextras-agent/src/main/kotlin/com/nicologies/prextras/BuildService.kt
package com.nicologies.prextras

import com.nicologies.prextras.common.PrExtrasConstants
import com.nicologies.prextras.common.SettingsKeys
import jetbrains.buildServer.RunBuildException
import jetbrains.buildServer.agent.AgentRunningBuild
import jetbrains.buildServer.agent.runner.BuildServiceAdapter
import jetbrains.buildServer.agent.runner.ProgramCommandLine
import jetbrains.buildServer.util.StringUtil
import org.eclipse.egit.github.core.*
import org.eclipse.egit.github.core.client.GitHubClient
import org.eclipse.egit.github.core.service.GitHubService
import org.eclipse.egit.github.core.service.IssueService
import org.eclipse.egit.github.core.service.PullRequestService

import java.io.IOException
import java.net.URL
import java.security.InvalidParameterException
import java.util.*

class BuildService : BuildServiceAdapter() {
    private var _isOnWindows = true

    private val gitHubApiUrl: URL
        get() {
            val url = GitHubApiUrlHelper.getURL(this.systemProperties)
            logger.message("GitHub API URL: " + url.toString())
            return url
        }

    @Throws(RunBuildException::class)
    override fun beforeProcessStarted() {
        logger.progressMessage("Running PrExtras")
        val runnerParams = runnerParameters
        val configParams = configParameters

        _isOnWindows = configParams["teamcity.agent.jvm.os.name"]!!.toLowerCase().startsWith("win")

        try {
            val build = build
            val repoUrl = getRepoUrl(configParams)
            val githubApiUrl = gitHubApiUrl

            val service = PullRequestService(GitHubClient(githubApiUrl.host,
                    githubApiUrl.port, githubApiUrl.protocol))
            initService(service)
            val issueService = IssueService(service.client)

            val prNum = configParams["teamcity.build.branch"]!!
            var pullRequest: PullRequest? = null
            val isPullRequestBuild = StringUtil.isNumber(prNum)
            val repo = getRepositoryId(repoUrl)
            if (isPullRequestBuild) {
                pullRequest = getPullRequest(service, repo, prNum)
            }
            exportConfigParam(build, "teamcity.build.pull_req.is_pull_req",
                    if (isPullRequestBuild) "true" else "false")

            val headBranchName = getHeadBranchName(pullRequest, prNum)

            exportConfigParam(build, HeadBranchNameParamName, headBranchName)

            val baseBranchName = getBaseBranchName(pullRequest, prNum)

            exportConfigParam(build, BaseBranchNameParamName, baseBranchName)

            ExportPullRequestExtraInfo(pullRequest, build, configParams, issueService)

            if (pullRequest != null && pullRequest.comments > 0) {
                exportPrParticipants(service, repo, pullRequest.number)
            }

            val appendToBuildNum = runnerParams[SettingsKeys.AppendToBuildNum]
            if (!StringUtil.isEmptyOrSpaces(appendToBuildNum) && appendToBuildNum!!.toLowerCase() == "true") {
                var buildNum = configParameters["build.number"]!!
                val isGitVersionBuildNum = buildNum.matches("(?i).*PullRequest\\.\\d*\\+.*".toRegex())
                if (isGitVersionBuildNum) {
                    buildNum = buildNum.replace("(?i)PullRequest\\.\\d*".toRegex(), headBranchName)
                } else if (!buildNum.contains(headBranchName)) {
                    buildNum = buildNum + "-" + headBranchName
                }
                buildNum = buildNum.replace("(?i)PullRequest".toRegex(), "PR")
                val param = "##teamcity[buildNumber '$buildNum']"
                logger.message(param)
            }

            failBuildIfConflict(runnerParams, pullRequest)
        } catch (e: RunBuildException) {
            throw e
        } catch (e: Exception) {
            throw RunBuildException(e.message, e)
        }

    }

    private fun exportConfigParam(build: AgentRunningBuild, key: String, value: String?) {
        ParamExporter.exportConfigParam(build, key, value)
    }

    private fun ExportPullRequestExtraInfo(pullRequest: PullRequest?,
                                           build: AgentRunningBuild,
                                           configParams: Map<String, String>,
                                           issueService: IssueService) {
        if (pullRequest == null) {
            exportConfigParam(build, "teamcity.build.pull_req.is_merged", "false")
            exportConfigParam(build, "teamcity.build.pull_req.is_closed", "false")
            return
        }
        exportConfigParam(build, "teamcity.build.pull_req.is_merged", if (pullRequest.isMerged) "true" else "false")
        exportConfigParam(build, "teamcity.build.pull_req.is_closed",
                if (pullRequest.closedAt != null) "true" else "false")

        val user = pullRequest.user
        var email = user.email
        if (!StringUtil.isEmptyOrSpaces(email)) {
            exportConfigParam(build, "teamcity.build.pull_req.author_email", email)
        }

        val authorName = getGitHubUser(user)
        if (!StringUtil.isEmptyOrSpaces(authorName)) {
            exportConfigParam(build, "teamcity.build.pull_req.author", authorName)
        }

        val htmlUrl = pullRequest.htmlUrl
        if (!StringUtil.isEmptyOrSpaces(htmlUrl)) {
            exportConfigParam(build, "teamcity.build.pull_req.url", htmlUrl)
        }

        val assignee = pullRequest.assignee
        if (assignee != null) {
            val assigneeName = getGitHubUser(assignee)
            if (!StringUtil.isEmptyOrSpaces(assigneeName)) {
                exportConfigParam(build, "teamcity.build.pull_req.assignee", assigneeName)
            }
            email = assignee.email
            if (!StringUtil.isEmptyOrSpaces(email)) {
                exportConfigParam(build, "teamcity.build.pull_req.assignee_email", email)
            }
        }
        val repo = getRepositoryId(getRepoUrl(configParams))
        try {
            val issue = issueService.getIssue(repo, pullRequest.number)
            val labels = issue.labels
            if (!labels.isEmpty()) {
                val labelsString = StringUtil.join(";", *labels.toTypedArray())
                exportConfigParam(build, "teamcity.build.pull_req.labels", labelsString)
            }
        } catch (e: IOException) {
            logger.message(e.message + e.stackTrace)
            throw RuntimeException(e.message, e)
        }

    }

    private fun exportPrParticipants(service: PullRequestService, repo: RepositoryId, prNum: Int) {
        try {
            val participants = HashSet<String>()
            getParticipantsFromComments(participants, service.getComments(repo, prNum))
            val issueService = IssueService(service.client)
            getParticipantsFromComments(participants, issueService.getComments(repo, prNum))
            if (participants.isEmpty()) {
                return
            }

            exportConfigParam(build, "teamcity.build.pull_req.participants",
                    StringUtil.join(";", participants))
        } catch (e: IOException) {
            e.printStackTrace()
            logger.error(e.message)
        }

    }

    private fun getParticipantsFromComments(participants: HashSet<String>, comments: List<Comment>) {
        comments
                .map { getGitHubUser(it.user) }
                .filterNot { StringUtil.isEmptyOrSpaces(it) }
                .filterNotNullTo(participants)
    }

    @Throws(RunBuildException::class)
    private fun failBuildIfConflict(runnerParams: Map<String, String>,
                                    pr: PullRequest?) {
        if (pr == null) {
            return
        }
        val failBuildIfConflict = runnerParams[SettingsKeys.FailBuildIfConflict]

        if (!StringUtil.isEmptyOrSpaces(failBuildIfConflict) && failBuildIfConflict.equals("true", ignoreCase = true)) {
            if (!pr.isMerged && !pr.isMergeable) {
                val msg = "The pull request has conflicts " + pr.htmlUrl
                throw RunBuildException(msg)
            }
        }
    }

    @Throws(RunBuildException::class)
    private fun getHeadBranchName(pr: PullRequest?, prNum: String): String {
        if (!StringUtil.isNumber(prNum)) {
            return prNum
        }
        try {
            return if (pr != null) pr.head.ref else prNum
        } catch (ex: Exception) {
            throw RunBuildException("Unable to get head branch name for pull request " + ex.message, ex)
        }

    }

    @Throws(RunBuildException::class)
    private fun getBaseBranchName(pr: PullRequest?, prNum: String): String {
        if (!StringUtil.isNumber(prNum)) {
            return prNum
        }
        try {
            return if (pr != null) pr.base.ref else prNum
        } catch (ex: Exception) {
            throw RunBuildException("Unable to get base branch name for pull request " + ex.message, ex)
        }

    }

    @Throws(RunBuildException::class)
    private fun getPullRequest(service: PullRequestService, repo: RepositoryId, prNum: String): PullRequest {
        val pr: PullRequest
        val prNumInteger: Int
        try {
            prNumInteger = Integer.parseInt(prNum)
        } catch (e: Exception) {
            throw RunBuildException(prNum + " is not pull request number: " + e.message, e)
        }

        logger.message("Trying to get extra information of pull request " + prNum + " from "
                + repo.owner + "'s " + repo.name)
        try {
            pr = service.getPullRequest(repo, prNumInteger)
        } catch (e: IOException) {
            throw RunBuildException("unable to get pull request info: " + e.message, e)
        }

        return pr
    }

    private fun getRepositoryId(repoUrl: String): RepositoryId {
        val id = RepoInfoParser.Parse(repoUrl)
        logger.message("owner: " + id.owner + " repo: " + id.name)
        return id
    }

    private fun getRepoUrl(configParams: Map<String, String>): String {
        return configParams["vcsroot.url"]!!
    }

    private fun initService(service: GitHubService) {
        val runnerParams = runnerParameters
        val authType = runnerParams[SettingsKeys.AuthType]
        if (authType == PrExtrasConstants.SystemWideTokenAuthType) {
            val token = systemProperties[SettingsKeys.GithubToken]
            if (StringUtil.isEmptyOrSpaces(token)) {
                throw InvalidParameterException("Please set the system wide github token")
            }
            service.client.setOAuth2Token(token)
        } else if (authType == PrExtrasConstants.TokenAuthType) {
            val token = runnerParams[SettingsKeys.GithubToken]
            if (StringUtil.isEmptyOrSpaces(token)) {
                throw InvalidParameterException("Please set the github token in the build step's settings page")
            }
            service.client.setOAuth2Token(token)
        } else {
            var password: String? = runnerParams[SettingsKeys.GithubPassword]
            if (StringUtil.isEmptyOrSpaces(password)) {
                password = null
            }
            val user = runnerParams[SettingsKeys.GithubUserName]
            if (StringUtil.isNotEmpty(user)) {
                service.client.setCredentials(user, password)
            }
        }
    }

    @Throws(RunBuildException::class)
    override fun makeProgramCommandLine(): ProgramCommandLine {
        return object : ProgramCommandLine {
            @Throws(RunBuildException::class)
            override fun getExecutablePath(): String {
                return if (_isOnWindows) "cmd" else "sh"
            }

            @Throws(RunBuildException::class)
            override fun getWorkingDirectory(): String {
                return checkoutDirectory.path
            }

            @Throws(RunBuildException::class)
            override fun getArguments(): List<String> {
                val ret = Vector<String>()
                if (_isOnWindows) {
                    ret.add("/c")
                } else {
                    ret.add("-c")
                }
                ret.add("echo")
                ret.add("Branch name is " + configParameters[HeadBranchNameParamName])
                return ret
            }

            @Throws(RunBuildException::class)
            override fun getEnvironment(): Map<String, String> {
                return buildParameters.environmentVariables
            }
        }
    }

    companion object {
        private val HeadBranchNameParamName = "teamcity.build.pull_req.branch_name"
        private val BaseBranchNameParamName = "teamcity.build.pull_req.base_branch_name"

        private fun getGitHubUser(user: User?): String? {
            if (user == null) {
                return null
            }
            val login = user.login
            return if (StringUtil.isEmptyOrSpaces(login)) {
                null
            } else login
        }
    }
}
